sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1106 @@
+"""
+The :mod:`sklearn.model_selection.split` module includes classes and
+functions to split the data based on a preset strategy.
+"""
+
+# Author: Alexandre Gramfort <alexandre.gramfort@inria.fr>,
+#         Gael Varoquaux <gael.varoquaux@normalesup.org>,
+#         Olivier Girsel <olivier.grisel@ensta.org>
+# License: BSD 3 clause
+
+
+from __future__ import print_function
+from __future__ import division
+
+import warnings
+from itertools import chain, combinations
+from math import ceil, floor, factorial
+import numbers
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+import scipy.sparse as sp
+
+from ..utils import indexable, check_random_state, safe_indexing
+from ..utils.validation import _num_samples, check_array, column_or_1d
+from ..utils.multiclass import type_of_target
+from ..externals.six import with_metaclass
+from ..externals.six.moves import zip
+from ..utils.fixes import bincount
+
+
+__all__ = ['KFold',
+           'LeaveOneLabelOut',
+           'LeaveOneOut',
+           'LeavePLabelOut',
+           'LeavePOut',
+           'ShuffleSplit',
+           'StratifiedKFold',
+           'StratifiedShuffleSplit',
+           'PredefinedSplit',
+           'train_test_split',
+           'check_cv']
+
+
+class _PartitionIterator(with_metaclass(ABCMeta)):
+    """Base class for CV iterators where train_mask = ~test_mask
+
+    Implementations must define `_iter_test_masks` or `_iter_test_indices`.
+    """
+
+    def split(self, X, y=None, labels=None):
+        """Generate train/test indices to split data in train test sets.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training vectors, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            Target values (class labels in classification, real numbers in
+            regression)
+
+        labels : array-like of int with shape (n_samples,), optional
+            Arbitrary domain-specific stratification of the data to be used
+            to draw the splits.
+        """
+        self._pre_split_check(X, y, labels)
+        ind = np.arange(_num_samples(X))
+        for test_index in self._iter_test_masks(X, y, labels):
+            train_index = ind[np.logical_not(test_index)]
+            test_index = ind[test_index]
+            yield train_index, test_index
+
+    @abstractmethod
```

Comment:
actually, I'm not entirely convinced that this is the way to go. You rarely use this function at all, right? Which Iterators actually need it?
Depending on how frequent this is actually implemented, instead we could do

```python
def split(self, X, y, labels=None):
    # checking here
    super(myclass, self).split(X, y, labels=labels)
```

in the estimators that actually do checking.